Mathlib/Tactic/Simpa.lean

```diff
@@ -0,0 +1,47 @@
+/-
+Copyright (c) 2018 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Arthur Paulino, Gabriel Ebner, Mario Carneiro
+-/
+import Lean
+import Mathlib.Tactic.Core
+
+namespace Mathlib.Tactic
+
+open Lean Parser.Tactic Elab.Tactic
+
+/--
+This is a "finishing" tactic modification of `simp`. It has two forms.
+
+* `simpa [rules, ⋯] using e` will simplify the goal and the type of
+  `e` using `rules`, then try to close the goal using `e`.
+
+  Simplifying the type of `e` makes it more likely to match the goal
+  (which has also been simplified). This construction also tends to be
+  more robust under changes to the simp lemma set.
+
+* `simpa [rules, ⋯]` will simplify the goal and the type of a
+  hypothesis `this` if present in the context, then try to close the goal using
+  the `assumption` tactic.
+
+#TODO: implement `with ⋯` behavior
+-/
+elab "simpa " cfg?:(config)? disch?:(discharger)?
+    only?:&" only "? args?:(simpArgs)?
+    wth?:(withArgs)? using?:(usingArg)? : tactic => do
+  let cfg := cfg?.getOptional?
+  let disch := disch?.getOptional?
+  let only := only?.getOptional?
+  let args ← args?.getOptional?.mapM getSimpArgs
+  let nGoals := (← getUnsolvedGoals).length
+  evalTactic $ ← `(tactic|simp $(cfg)? $(disch)? $[only%$only]? $[[$[$args],*]]?)
+  if (← getUnsolvedGoals).length < nGoals then
+    throwError "try 'simp' instead of 'simpa'"
+  match ← using?.getOptional?.mapM getUsingArg with
+  | none   =>
+    evalTactic $ ← `(tactic|try simp $(cfg)? $(disch)? $[only%$only]? $[[$[$args],*]]? at this)
```

Comment:
In general, I'm not a big fan of using `try` because it just hides errors and you have no idea what went wrong as a user.  But at least in this case I can't think of any error condition that would be swallowed by `try` but wouldn't show up in the first `simp`.